### PR TITLE
bazel: implement set-gen as a bazel genrule

### PR DESCRIPTION
**Pull request by mikedanese**
cc @caesarxuchao who is the gengo master until @lavalamp is back.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35733)

### CODE REVIEW

**Comment by spxtr** [comment]
That's pretty neat.

File: cmd/libs/go2idl/set-gen/BUILD [hunk 1]
```diff
@@ -20,3 +20,7 @@ go_binary(
         "//vendor:k8s.io/gengo/examples/set-gen/generators",
     ],
 )
+
+exports_files([
```
Comment (caesarxuchao) [inline cmd/libs/go2idl/set-gen/BUILD]: Can we use exports_files to replace importboss?

File: pkg/util/sets/BUILD [hunk 2]
```diff
@@ -13,14 +13,8 @@ load(
 go_library(
     name = "go_default_library",
     srcs = [
-        "byte.go",
-        "doc.go",
-        "empty.go",
-        "int.go",
-        "int64.go",
-        "string.go",
+        ":set-gen",
```
Comment (caesarxuchao) [inline pkg/util/sets/BUILD]: Does ":" mean using genrule rather than treat as source file?

**Comment by caesarxuchao** [comment]
LGTM.

File: pkg/util/sets/BUILD (hunk 2 above)
Comment (mikedanese) [inline pkg/util/sets/BUILD]: This means use the output of the genrule as the source. :set-gen is the label of the set-gen genrule.

Comment (caesarxuchao) [inline pkg/util/sets/BUILD]: I see. Thanks.

**Comment by ixdy** [comment]
what's the legacy way of doing this outside bazel?

**Comment by mikedanese** [comment]
https://github.com/kubernetes/kubernetes/blob/master/hack/update-codegen.sh does this along with other stuff

**Pull request merged**